### PR TITLE
`Development`: Enable Aeolus for Jenkins

### PR DESCRIPTION
**Pull request by reschandreas**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation, Description, and Context

This PR adds the platform-independent CI configuration Aeolus to the Jenkins service within Aeolus. This does not change any existing editing features or diminish the status quo of the Jenkins integration. It simply adds a new possible modus operandi to Artemis that does not require the Jenkinsfiles to be written in a general way and reduced rest calls to Artemis during their execution as the Jenkinsfile directly contains the definition.

### Steps for Testing

Prerequisites:
- local artemis setup with jenkins and gitlab

1. Log in to Artemis
2. Create a Programming Exercise with Aeolus enabled and check if your changes were generated correctly

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/programming-exercise.model.ts.html) | 100% | ✅ |
| [programming-exercise-custom-aeolus-build-plan.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/custom-build-plans/programming-exercise-custom-aeolus-build-plan.component.ts.html) | 91.56% | ✅|
| [aeolus.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/service/aeolus.service.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ProgrammingExerciseStudentParticipationRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/ProgrammingExerciseStudentParticipationRepository.html) | 80% | ✅ |
| [ProfileService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ProfileService.html) | 100% | ✅ |
| [Action.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.aeolus/Action.html) | 100% | ✅ |
| [AeolusBuildPlanService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.aeolus/AeolusBuildPlanService.html) | 97% | ✅ |
| [AeolusTemplateService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.aeolus/AeolusTemplateService.html) | 88% | ✅ |
| [PlatformAction.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.aeolus/PlatformAction.html) | 100% | ✅ |
| [Windfile.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.aeolus/Windfile.html) | 100% | ✅ |
| [WindfileMetadata.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.aeolus/WindfileMetadata.html) | 100% | ✅ |
| [BambooBuildPlanService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.bamboo/BambooBuildPlanService.html) | 87% | ✅ |
| [JenkinsBuildPlanService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.jenkins.build_plan/JenkinsBuildPlanService.html) | 95% | ✅ |
| [ProgrammingExerciseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5967/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseService.html) | 91% | ✅ |

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to unset `buildPlanId` for exercises, enhancing exercise management.
	- Added support for conditional creation of custom Aeolus build plans in exercises.
- **Enhancements**
	- Expanded the `ProgrammingExerciseService` constructor to support additional services.
	- Improved handling of tokens and headers in build plan services for better security and integration.
- **Refactor**
	- Streamlined imports across various services for cleaner code.
	- Updated method signatures and internal logic for clarity and efficiency.
- **Tests**
	- Added tests for handling failed build plan publishing and unauthorized REST calls.
	- Expanded test coverage for Aeolus service functionalities.

### CODE REVIEW

**Comment by konrad2002** [comment]
if this pr is meant to be tested on test servers, could you please provide a more detailed testing description? How do we create a "Jenkins Programming Exercise" (is it just a programming exercise on a test server with Jenkins?) and how to enable Aeolus? (I tried hard to find any options about it, but maybe it is just my stupidity :DD)
If this pr can not be tested on the test server, pls make this clear in the testing steps and instructions.

**Comment by reschandreas** [comment]
@konrad2002 yes, sorry. Right now aeolus is not active on any test servers, I'm planning on hosting a testing session soon where I will present the changes

**Review by rstief — DISMISSED** [review]
Demonstrated locally, looked good 👍

**Review by milljoniaer — DISMISSED** [review]
Demonstrated in testing session, works as expected

**Review by milljoniaer — DISMISSED** [review]
reapprove after code changes

**Review by laurenzfb — APPROVED** [review]
code lgtm. demonstrated in testing session, works

**Review by rstief — DISMISSED** [review]
Reapproved after changes

**Review by milljoniaer — DISMISSED** [review]
Reapprove

**Review by b-fein — COMMENTED** [review]
Aelous ‘Add action’ does nothing for Java+DejaGnu exercises. Maybe it should be disabled for this kind of exercises?
image

**Review by b-fein — COMMENTED** [review]
Tested with test server at Uni Passau: Basic setup of Maven and Gradle build plans works.

Noted problems:
- (optional enhancement) `mvn` commands should be replaced with `mvn -B` to disable the download progress
- Enabling static code analysis breaks the build plan: The Jenkins plugin that parses the test and SCA results expects the SCA ones to be in `staticCodeAnalysisResults/` rather than `results/`. This breaks now since e.g., the Checkstyle report is attempted to be parsed as a JUnit result.
- When editing the exercise, the ‘Customise build plan with Aeolus’ checkbox is unchecked again. When clicking it again, though, any manual customisations made are still there. Maybe the selection could be persisted?

**Comment by reschandreas** [comment]
Hi @b-fein,
thanks for testing this PR.
To the problems you noted, the first one I need to test if that still works for bamboo and localci. The second one is fixed on the aeolus part but I would like to address that in a followup PR if that is alright with you, same with the template for Java+DejaGnu exercises. I further always uncheck the Customise build plan with Aeolus as to signify that the current editing process did not further change it, but if that's unintuitive I'd also change that in a followup.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 1]
```diff
@@ -234,8 +242,8 @@ public ProgrammingExercise createProgrammingExercise(ProgrammingExercise program
         // make sure that plagiarism detection config does not use existing id
         Optional.ofNullable(programmingExercise.getPlagiarismDetectionConfig()).ifPresent(it -> it.setId(null));
 
-        // for LocalCI and Aeolus, we store the build plan definition in the database as a windfile
-        if (aeolusTemplateService.isPresent() && programmingExercise.getBuildPlanConfiguration() == null) {
+        // for LocalCI and Aeolus, we store the build plan definition in the database as a windfile, we don't do that for Jenkins
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:245]: Why don’t we do that for Jenkins? Some short explanation in the comment here would be helpful for others.

**Review by b-fein — DISMISSED** [review]
Sorry for not making clear that the noted problems are not required for the base setup in this PR but are of course fine for future improvements.

Code+Manual test: :+1:

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java (hunk 1 above)
Comment (reschandreas) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:245]: For Jenkins and Bamboo we only store the windfile if needed, in the case of Localci we always need it, I'll clarify this in a followup along with the other improvements.

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseStudentParticipationRepository.java [hunk 2]
```diff
@@ -193,4 +193,13 @@ default ProgrammingExerciseStudentParticipation findWithTeamStudentsByIdElseThro
                 WHERE p.buildPlanId IS NOT NULL or p.repositoryUri IS NOT NULL
             """)
     Page<ProgrammingExerciseStudentParticipation> findAllWithRepositoryUriOrBuildPlanId(Pageable pageable);
+
+    @Transactional
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseStudentParticipationRepository.java:197]: please add the usual comment that this use of @Transactional is ok

**Review by laurenzfb — DISMISSED** [review]
reapprove after adding tests

**Review by rstief — APPROVED** [review]
reapproved after changes

**Review by milljoniaer — APPROVED** [review]
reapprove after code changes

**Review by rstief — DISMISSED** [review]
Code also LGTM.

**Review by b-fein — APPROVED** [review]

**Review by b-fein — APPROVED** [review]

**Review by rstief — APPROVED** [review]
reapprove after changes

**Review by laurenzfb — APPROVED** [review]
reapprove

**Review by milljoniaer — APPROVED** [review]
reapprove after code changes

**Pull request merged**